app/controllers/users/webauthn_setup_controller.rb

```diff
@@ -108,6 +109,20 @@ def show_delete
 
     private
 
+    def confirm_already_set_ft
+      form = WebauthnVisitForm.new(
+        user: current_user,
+        url_options: url_options,
+        in_mfa_selection_flow: in_multi_mfa_selection_flow?,
+      )
+      result = form.submit(new_params)
+      @platform_authenticator = form.platform_authenticator?
+      if @platform_authenticator && in_account_creation_flow? && result &&
+         current_user.webauthn_configurations.platform_authenticators.present?
+        redirect_to authentication_methods_setup_path
+      end
+    end
```

Comment:
@aduth I see. I have moved that to a before_action but I'm not confident on how I executed it. Specifically around detecting that the user is trying to set another platform_authenticator. 
For some reason `result` doesn't appear to be meaningfully used but the redirect doesn't happen without that value being set. Can you see why it wouldn't work without that? TBH it's only added to the `if` statement because lint will rightfully complain.

Reply:
Yeah, I think the way this is set up with the form makes it a little awkward, with how `submit` needs to be called for the parameters to be "consumed". I think we could go one of two ways with this:

- Keep it largely as-is, maybe refactoring `form` and/or `result` to a new method that could be used throughout this file
- Bypass `form` and look at the parameters directly

I imagine the first would look something like:

```suggestion
    def validate_existing_platform_authenticator
      @visit_form.submit(new_params)
      if visit_form.platform_authenticator? && in_account_creation_flow? &&
         current_user.webauthn_configurations.platform_authenticators.present?
        redirect_to authentication_methods_setup_path
      end
    end

    def visit_form
      @visit_form ||= WebauthnVisitForm.new(
        user: current_user,
        url_options: url_options,
        in_mfa_selection_flow: in_multi_mfa_selection_flow?,
      )
    end
```

The second looking something like...

```suggestion
    def validate_existing_platform_authenticator
      if platform_authenticator? && in_account_creation_flow? &&
         current_user.webauthn_configurations.platform_authenticators.present?
        redirect_to authentication_methods_setup_path
      end
    end

    def platform_authenticator?
      params[:platform] == 'true'
    end
```

Couple notes on the above:

- I tend to see a convention with `before_action` often being prefixed with "validate_"
- Similarly, I tend to see our code refer to "face touch" as "platform authenticator", and generally I shy away from abbreviations that aren't immediately obvious to those not in-the-know
- We probably don't need to set the `@platform_authenticator` from your original proposal, since that's only relevant if it gets as far as calling the `new` action (which assigns it anyways)
- Similarly, I'm not sure there's ever a case that `result` would be `nil` for it to be relevant in the condition?

Reply:
`def platform_authenticator?
      params[:platform] == 'true'
    end
`

Yeah I was hoping something succinct like this could be leveraged.
I was wondering about how that sort of authentication is referred to as platform - is that to keep it sort of broadly termed to allow for other kinds of platform authentication in the future? Not strictly fingerprint/face recognition? Just curious.

Reply:
Yeah, it stems back to how the technical implementation maps to the specification's concept of [platform authenticators](https://www.w3.org/TR/webauthn-3/#platform-authenticators). And it does give us some flexibility in how we choose the user-facing language (what we currently call "Face or touch unlock").